Add code for set EPS configuration parameter

DavidGhergut opened the pull request:
I have added the `set_config_params` function with comments! Solves issue #37

alexbolfa committed:
Add output bus group off command

Add the "output bus group off command" (ICD page 41).

Issue: https://github.com/davincisatellite/msp-flightsoftware/issues/5

alexbolfa committed:
Fix output bus group off command CC

alexbolfa committed:
Fix fix output bus group off CC

alexbolfa committed:
Add output bus group state command

Add the "output bus group state command" (ICD page 42).

Issue: https://github.com/davincisatellite/msp-flightsoftware/issues/5

alexbolfa committed:
Add output bus channel on command

Add the "output bus group state command" (ICD page 43).

Issue: https://github.com/davincisatellite/msp-flightsoftware/issues/5

alexbolfa committed:
Add output bus channel off command

Add the "output bus channel off command" (ICD page 44).

Issue: https://github.com/davincisatellite/msp-flightsoftware/issues/5

alexbolfa committed:
Add switch to nominal mode command

Add the "switch to nominal mode" (ICD page 45).

Issue: https://github.com/davincisatellite/msp-flightsoftware/issues/5

alexbolfa committed:
Add switch to safety mode command

Add the "switch to safety mode" (ICD page 46).

Issue: https://github.com/davincisatellite/msp-flightsoftware/issues/5

alexbolfa committed:
Add get PDU/PIU overcurrent fault state command

Add the "PDU/PIU overcurrent fault state" (ICD page 48).

Issue: https://github.com/davincisatellite/msp-flightsoftware/issues/5

alexbolfa committed:
Add get reset configuration parameter command

Add the "reset configuration parameter" (ICD page 72).

Issue: https://github.com/davincisatellite/msp-flightsoftware/issues/5

alexbolfa committed:
Add Get PCU Housekeeping Data (RAW) command

Add the "Get PCU Housekeeping Data (RAW)" (ICD page 64).

Issue: https://github.com/davincisatellite/msp-flightsoftware/issues/5

alexbolfa committed:
Add Get PCU Housekeeping Data (ENG) command

Add the "Get PCU Housekeeping Data (ENG)" (ICD page 66).

Issue: https://github.com/davincisatellite/msp-flightsoftware/issues/5

Dennisman219 committed:
Removed empty define which broke building

alexbolfa committed:
Cast return value of reserved byte to void

Commit:
Merge branch 'fix/eps/remove-empty-define' into test/eps

Commit:
Merge remote-tracking branch origin/get-pcu-housekeeping-data-eng-command into test/eps

Dennisman219 committed:
xMerge remote-tracking branch origin/get-pcu-housekeeping-data-raw-command into test/eps

Dennisman219 committed:
Merge remote-tracking branch origin/get-pdu-overcurrent-fault-state into test/eps

Dennisman219 committed:
Added .vscode folder to gitignore

Dennisman219 committed:
Merge remote-tracking branch origin/add-switch-safety-mode into test/eps

Dennisman219 committed:
Merge remote-tracking branch origin/add-switch-nominal-mode into test/eps

Dennisman219 committed:
Merge remote-tracking branch origin/add-output-bus-channel-off into test/eps

Dennisman219 committed:
Merge remote-tracking branch origin/add-output-bus-channel-on into test/eps

Dennisman219 committed:
Merge remote-tracking branch origin/add-output-bus-group-state into test/eps

Dennisman219 committed:
Merge remote-tracking branch origin/add-output-bus-group-off into test/eps

Dennisman219 committed:
Merge remote-tracking branch origin/add-output-bus-group-off into test/eps

Dennisman219 committed:
removed double declaration for output_bus_channel_on command, must be reimplemented using ICD p.36

Dennisman219 committed:
removed double declaration of pcu_housekeeping_data_reply struct

Dennisman219 committed:
remove overloading watchdog command with pbu return type

Dennisman219 committed:
fix output_bus_group_state and output_bus_group_off bitfield passing

Dennisman219 committed:
manually fixed and added output-bus-group-on func/cmd by alex

Dennisman219 committed:
manually added add_reset_configuration from alex

DavidGhergut committed:
Reset configuration command

DavidGhergut committed:
Fixed conflict for add-configure-eps-parameters merging into test/eps branch

ViciousDoormat committed:
FIX: remove zeroes before numbers

Dennisman219 committed:
temp fix for msp.h include

DavidGhergut committed:
Ïnclude fixes for code composer studio

DavidGhergut committed:
Add code for configuring EPS parameters; solves issue #35

Dennisman219 committed:
Merge pull request #36 from davincisatellite/configure_EPS_parameters

Add code for configuring EPS parameters; solves issue #35

DavidGhergut committed:
Add code for configuring EPS parameters; solves issue #37

DavidGhergut committed:
Add set_config_params function header to EPS.h

mavaneerden commented on src/EPS.cpp:
Check error

ViciousDoormat replied:
What error?

mavaneerden replied:
Endtransmission can return an error code, though you dont actually need to use it because requestFrom already does that for you. Difference is that endtransmission sends a stop signal, and requestfrom does not. So you have to check what you need here

mavaneerden commented on src/EPS.cpp:
is this delay really necessary? It isn't done anywhere else where we use I2C.

DavidGhergut replied:
![image](https://github.com/user-attachments/assets/294c9611-5148-4ed7-9461-9b63d358766c)

According to this screenshot from Matteo, the delay is necessary, however we are using 25 ms instead of 20 ms as mentioned, but I think we can tweak this value in testing as well!

mavaneerden replied:
Okay, but it says here that you should probably check the CC/RC byte, and you are not doing that.

The fear I have is that you are blocking the entire OBC for 25ms. I am not sure how much of a problem that is, you should investigate.

ViciousDoormat replied:
I agree, we should investigate this, but that can probably only be done later in the project right? When parts actually request things from the EPS. 
BTW, as far as I understood their advice of checking the CC/RC in scheme 1, they advise checking the value after the period of 20ms. So this would not solve the issue you mention. What would potentially solve it is choosing scheme 2

mavaneerden replied:
yes the checking is a separate problem. If you want to investigate the delay later, make sure you have documented this somewhere (in the code preferably, and maybe somewehere else too).

ViciousDoormat replied:
Added the TODO to the code documentation

mavaneerden commented on src/EPS.cpp:
Check for nullptr first

ViciousDoormat replied:
Done

mavaneerden commented on src/EPS.cpp:
Have more error codes than just 0 or 1. We need to determine what exactly goes wrong. Use a different code for different failure modes, and don't use magic numbers but use enum classes instead.

ViciousDoormat replied:
I agree that this has to be determined. However, I do not think this is relevant to this specific issue. I will make a new issue for it and link to it in this comment

#52

mavaneerden commented on src/EPS.cpp:
I think this part is repeated in multiple other functions, perhaps you can isolate it so it can be reused without copy-paste.

mavaneerden commented on src/EPS.cpp:
Don't use magic numbers here.

DavidGhergut replied:
@mavaneerden I completely agree here! There is an issue regarding refactoring already up, see #34 where creating constants for these values is one of the refactoring points and it has been addressed in pull request #40 !

DavidGhergut committed:
Fix bug 2, 3 and 4; issue #33; Create constants for common values in the code; issue #34

DavidGhergut committed:
Remove unnecessary comments and add ICD references for STID, IVID and BID

DavidGhergut committed:
Add writeCommand function and replace the repeating code with function calls

DavidGhergut committed:
Replace consts with enums

DavidGhergut committed:
Add comments for reply attributes

ViciousDoormat committed:
Merge pull request #40 from davincisatellite/refactoring_bug_fixes_untested_features

Fix bug 2, 3 and 4 and create constants for common values in the code

mavaneerden commented on src/EPS.cpp:
```diff
@@ -58,6 +92,89 @@ uint8_t get_param_length(uint16_t par_id) {
     }
 }
 
+void writeCommand(Dwire &wire, uint8_t i2c_address, Identifiers stid, Identifiers ivid, CommandCode commandCode, Identifiers bid) {
```
STID, IVID and BID are always the same, I think you should remove these from the parameters

mavaneerden commented on src/EPS.cpp:
```diff
@@ -3,6 +3,39 @@
+    BID = 0x00                                        // reference: page 20 of 87 (ICD)
+};
+
+enum CommandCode {
```
If you use an `enum class` here, you don't need to prefix your enums with `CMD_CODE`, you can just use `Command::MY_COMMAND`

mavaneerden commented on src/EPS.cpp:
```diff
@@ -3,6 +3,39 @@
+    CONF_KEY_RESET_CONFIGURATION = 0xA7               // reference: page 66 of 87 (ICD)
+};
+
 bool is_valid_param_id(uint16_t par_id) {
```
I think if you make Parameter IDs an enum class, you don't even need this validation function because it'll check if it is a Parameter ID instance at compile time. You should still keep the length check though, or alternatively make the compiler error when not all possible enum values have been used in a switch case

ViciousDoormat replied:
Done

mavaneerden commented on src/EPS.cpp:
Would be nice to add this piece to the wireCommand function, since it is used elsewhere as well

ViciousDoormat replied:
First of all, after going through the code I could not find another part that does this as well. However, I think there will be added one additional command that will do it.
Second, how do you envision this? The command writes other stuff between that for loop and the write command, and most commands do not use this for loop.
Maybe it would be better to make it its own function

mavaneerden replied:
Yeah in hindsight, you can leave it like this

ViciousDoormat committed:
FEAT: Add PCU housekeeping data running average command

Note that this does not yet contain the proposed refactoring of the reading functionality into a command, and this does not fix the potential issue of bit-shifting to the left by 1. Those are separate issues.

ViciousDoormat committed:
Merge pull request #47 from davincisatellite/eps/feature/46/houskeeping-data-average

FEAT: Add PCU housekeeping data running average command

ViciousDoormat committed:
FEAT: Get PBU housekeeping data (rolling average)

Note that this does not yet contain the proposed refactoring of the reading functionality into a command, and this does not fix the potential issue of bit-shifting to the left by 1. Those are separate issues.

ViciousDoormat committed:
Merge pull request #49 from davincisatellite/eps/feature/48/get-pbu-housekeeping-data-rolling-average

FEAT: Get PBU housekeeping data (rolling average)

mavaneerden commented on src/EPS.h:
```diff
@@ -121,6 +121,7 @@ class EPS {
     static standard_reply output_bus_group_on(DWire &wire, uint8_t i2c_address, uint16_t bitflag);
     static standard_reply reset_configuration(DWire &wire, uint8_t i2c_address);
     static config_reply reset_config_params(DWire &wire, uint8_t i2c_address, uint16_t par_id);
+    static config_reply set_config_params(DWire &wire, uint8_t i2c_address, uint16_t par_id, uint8_t *par_val);
```
Why are all the functions here static?

ViciousDoormat replied:
That's a great question I never actually asked haha. 
I think because the driver can be seen as something without a state, and I think there are multiple parts in the project that need access to it and this gives them easy access.
If you would make it a class, it would have to be a singleton (I think). You could argue this would be better, but due to the lack of state I do not think this is such a problem

mavaneerden replied:
Why is it stateless? Shouldnt it have a dedicated dwire and i2c bus as its state?
Because now you have to pass the same i2c address and dwire bus to the driver every time, which is tedious and error-prone

ViciousDoormat replied:
We can discuss this during next meeting. I think it is unwise to put effort into this refactoring for a state that would be that small

ViciousDoormat committed:
REF/BUG: Add ReplyBase and readCommand and refactor accordingly

ViciousDoormat committed:
BUG: Add dependency

ViciousDoormat committed:
BUG: Fix minor bugs

ViciousDoormat committed:
BUG: revert uint16 back to uint8

ViciousDoormat committed:
Merge pull request #51 from davincisatellite/eps/fix/50/refactorings-and-bug-fix

REF/BUG: Add ReplyBase and readCommand and refactor accordingly

ViciousDoormat committed:
REFACTOR: Introduce ParameterID enum

ViciousDoormat committed:
BUG: Add check for null pointer

ViciousDoormat commented:
Currently, also with the ParameterID enum I introduced, we only capture the data types the paramteres can have. They are based on table 3-24 on page 77. However, the next table showcases all paramters, a lot more than we currently capture. They are categorized based on the data type codes in tabel 77 (all parameters that give a uint16 start with 0x4). We need to find a nice way to capture this complexity

mavaneerden commented:
> Currently, also with the ParameterID enum I introduced, we only capture the data types the paramteres can have. They are based on table 3-24 on page 77. However, the next table showcases all paramters, a lot more than we currently capture. They are categorized based on the data type codes in tabel 77 (all parameters that give a uint16 start with 0x4). We need to find a nice way to capture this complexity

That is because the ParameterID enum doesnt actually contain parameter IDs, but parameter types. These are separate things

ViciousDoormat commented:
> > Currently, also with the ParameterID enum I introduced, we only capture the data types the paramteres can have. They are based on table 3-24 on page 77. However, the next table showcases all paramters, a lot more than we currently capture. They are categorized based on the data type codes in tabel 77 (all parameters that give a uint16 start with 0x4). We need to find a nice way to capture this complexity
> 
> That is because the ParameterID enum doesnt actually contain parameter IDs, but parameter types. These are separate things

Exactly, that is my point. But just making a huge enum is unwise I think, because parts are clearly grouped by type. I referred to this when I said we need a nice way to capture the complexity. 
Maybe the best solution would be an enum per type. However, then the parameter passed to the set conf param function would need to encompass all those enums. I think the union type we discussed during the meeting might be able to do that.

mavaneerden commented:
> > > Currently, also with the ParameterID enum I introduced, we only capture the data types the paramteres can have. They are based on table 3-24 on page 77. However, the next table showcases all paramters, a lot more than we currently capture. They are categorized based on the data type codes in tabel 77 (all parameters that give a uint16 start with 0x4). We need to find a nice way to capture this complexity
> > 
> > 
> > That is because the ParameterID enum doesnt actually contain parameter IDs, but parameter types. These are separate things
> 
> Exactly, that is my point. But just making a huge enum is unwise I think, because parts are clearly grouped by type. I referred to this when I said we need a nice way to capture the complexity. Maybe the best solution would be an enum per type. However, then the parameter passed to the set conf param function would need to encompass all those enums. I think the union type we discussed during the meeting might be able to do that.

I don't see why using a giant enum is problematic. You can put all parameters in there, this allows the higher layer to forget about the exact parameter code, and allows you to check valid IDs at compile time. As for checking the type, you mentioned that the parameter ID starts with a certain number, I think you can just check that number to find the type.

ViciousDoormat commented:
> > > > Currently, also with the ParameterID enum I introduced, we only capture the data types the paramteres can have. They are based on table 3-24 on page 77. However, the next table showcases all paramters, a lot more than we currently capture. They are categorized based on the data type codes in tabel 77 (all parameters that give a uint16 start with 0x4). We need to find a nice way to capture this complexity
> > > 
> > > 
> > > That is because the ParameterID enum doesnt actually contain parameter IDs, but parameter types. These are separate things
> > 
> > 
> > Exactly, that is my point. But just making a huge enum is unwise I think, because parts are clearly grouped by type. I referred to this when I said we need a nice way to capture the complexity. Maybe the best solution would be an enum per type. However, then the parameter passed to the set conf param function would need to encompass all those enums. I think the union type we discussed during the meeting might be able to do that.
> 
> I don't see why using a giant enum is problematic. You can put all parameters in there, this allows the higher layer to forget about the exact parameter code, and allows you to check valid IDs at compile time. As for checking the type, you mentioned that the parameter ID starts with a certain number, I think you can just check that number to find the type.

If we go for a giant enum, the functions calling the get/set-param commands do not have to explicitly use the type of the value they get/set. This seems very error-prone.

ViciousDoormat committed:
FEAT: pdu data commands + pbu eng + refactoring of data commands

mavaneerden commented:
> > > > > Currently, also with the ParameterID enum I introduced, we only capture the data types the paramteres can have. They are based on table 3-24 on page 77. However, the next table showcases all paramters, a lot more than we currently capture. They are categorized based on the data type codes in tabel 77 (all parameters that give a uint16 start with 0x4). We need to find a nice way to capture this complexity
> > > > 
> > > > 
> > > > That is because the ParameterID enum doesnt actually contain parameter IDs, but parameter types. These are separate things
> > > 
> > > 
> > > Exactly, that is my point. But just making a huge enum is unwise I think, because parts are clearly grouped by type. I referred to this when I said we need a nice way to capture the complexity. Maybe the best solution would be an enum per type. However, then the parameter passed to the set conf param function would need to encompass all those enums. I think the union type we discussed during the meeting might be able to do that.
> > 
> > 
> > I don't see why using a giant enum is problematic. You can put all parameters in there, this allows the higher layer to forget about the exact parameter code, and allows you to check valid IDs at compile time. As for checking the type, you mentioned that the parameter ID starts with a certain number, I think you can just check that number to find the type.
> 
> If we go for a giant enum, the functions calling the get/set-param commands do not have to explicitly use the type of the value they get/set. This seems very error-prone.

That is true, but you need an ID *and* a value somehow, they can use separate structures I think. But I can't really think of a way to do that nicely.

ViciousDoormat committed:
Merge pull request #54 from davincisatellite/eps/feature/53/pdu-and-pbu-data-commands

FEAT: pdu data commands + pbu eng + refactoring of data commands

ViciousDoormat committed:
DOC: Added delay investigation TODO

ViciousDoormat committed:
FEAT: Config param commands

DavidGhergut committed:
Merge pull request #55 from davincisatellite/eps/feature/config-param-commands

FEAT: Config param commands

DavidGhergut reviewed (commented):
I think we can merge now since we have a lot of changes for a single PR :)!
